### PR TITLE
Concurrent editing notifications - handle "is editing" and "latest revision" states

### LINKED ISSUE #12079: Concurrent editing notifications - showing users with unsaved changes
Part of [RFC 95 - concurrent editing notifications](https://github.com/wagtail/rfcs/pull/95)

As per the RFC section 'Distinguishing between "currently viewing" and "currently editing"', extend the EditingSession model, ping endpoint and front-end to indicate editors who have unsaved changes in their browser, as determined by the "dirty changes" mechanism.

(As part of this, evaluate whether the dirty changes mechanism works well enough for this purpose, and whether any fixes need to be applied.)

### Working on this

To be implemented by the Wagtail product team at Torchbox - we are not seeking external contributors for this.

### LINKED ISSUE #12080: Concurrent editing notifications - backend additions to identify concurrent edits
Part of [RFC 95 - concurrent editing notifications](https://github.com/wagtail/rfcs/pull/95)

As per the RFC section 'Notifying users on changes', add support to the 'ping' endpoint for passing a revision ID so that any users who have created revisions newer than that one are annotated as part of the response, including users who do not have an active editing session open.

(TBC: does this backend code need to know about the list of dismissed revision IDs, or is this entirely a front-end detail? If the former, that should be covered as part of this issue.)

### Working on this

To be implemented by the Wagtail product team at Torchbox - we are not seeking external contributors for this.

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #12079 and fixes #12080. Supersedes #12106. Addresses the comments in that PR and added more tests to cover the different edge cases. Notable changes in behaviour:

- Only the session of the user who created the latest revision is annotated with `revision_id`. This is to avoid showing too many "warning" signs and red outlined avatars. We only care about the latest version anyway.
- Only show the current user's "other session" if there's a different session that has `is_editing = True` or if the latest revision (that's not the one in hand) is created by the current user.
- Sort the final "other sessions" list based on the following logic:
  1. Prioritise any session with the latest revision. Then,
  2. Prioritise any session that is currently editing. Then,
  3. Prioritise any session with the smallest id, so that new sessions are appended to the end of the list (they're shown last). We are not sorting by last_seen_at to avoid shifting the order of the sessions as they ping the server.

### CODE REVIEW

File: wagtail/admin/views/editing_sessions.py [hunk 1]
```diff
@@ -100,27 +100,28 @@ def ping(request, app_label, model_name, object_id, session_id):
...
+        newest_revision = (
             all_revisions.filter(created_at__gt=original_revision.created_at)
-            .order_by("created_at")
+            .order_by("-created_at", "-pk")
             .select_related("user")
+            .first()
         )
 
-        for revision in newer_revisions:
+        if newest_revision and newest_revision.id != revision_id:
```
[the remark this answers is not in the document]
Comment (laymonage) [reply wagtail/admin/views/editing_sessions.py]: Good point, I was considering a case where the original revision's `created_at` was somehow modified on the database before this query is performed. But I think it's safe to say that it is an extremely unlikely race condition to happen 😆

File: wagtail/admin/views/editing_sessions.py [hunk 2]
```diff
@@ -123,6 +123,28 @@ def ping(request, app_label, model_name, object_id, session_id):
...
+    other_sessions = sorted(
+        other_sessions,
+        key=lambda other_session: other_session["is_editing"],
+        reverse=True,
+    )
+    other_sessions = sorted(
+        other_sessions,
+        key=lambda other_session: other_session["revision_id"] or 0,
+        reverse=True,
+    )
```
[the remark this answers is not in the document]
Comment (laymonage) [reply wagtail/admin/views/editing_sessions.py:159]: We need to sort the `revision_id` and `is_editing` in descending order, but `session_id` in ascending order, so I followed the [multi-step technique from Python docs](https://docs.python.org/3/howto/sorting.html#sort-stability-and-complex-sorts). But I suppose since there can only be one session (at most) with `revision_id`, we only care about the presence and not the value itself, so we can coerce this to a boolean and negate both values to sort in descending order.

I've addressed this in [`714dc6f` (#12109)](https://github.com/wagtail/wagtail/pull/12109/commits/714dc6f0729cfb200d230667dcd4e54e6a74ab56) (along with the other comment for the `if` check).

**Review by gasman — APPROVED** [review]
LGTM!
